ambry-mysql/src/main/java/com/github/ambry/mysql/BatchUpdater.java

```diff
@@ -96,58 +96,63 @@ public BatchUpdater(MySqlDataAccessor dataAccessor, String sql, String tableName
    */
   protected synchronized void addUpdateToBatch(GenericThrowableConsumer<PreparedStatement, SQLException> valueSupplier)
       throws SQLException {
+    boolean failed = false;
     try {
       if (startTime == 0) {
         startTime = System.currentTimeMillis();
       }
       if (maxBatchSize != 0 && currentBatchSize >= maxBatchSize) {
-        executeBatch();
+        executeBatchAndCommit();
       }
       valueSupplier.accept(statement);
       statement.addBatch();
       currentBatchSize++;
-      totalBatchSize++;
     } catch (SQLException e) {
-      dataAccessor.onException(e, Write);
-      logger.error("Failed to add batch on {}", tableName, e);
+      failed = true;
+      dataAccessor.onException(e, BatchUpdate);
+      logger.error("Failed to add batch on {}, already committed {}, rolling back", tableName, committedBatchSize, e);
+      dataAccessor.rollback();
       throw e;
+    } finally {
+      if (failed) {
+        dataAccessor.setAutoCommit(autoCommit);
```

Comment:
Why only if failed?

Reply:
Moving setAutoCommit to rollback method. This way the code looks a bit cleaner.